src/tracer.zig

```diff
@@ -0,0 +1,332 @@
+//! The tracer records a tree of event spans.
+//!
+//! In order to create event spans, you need somewhere to store the `SpanStart`.
+//!
+//!     var slot: ?SpanStart = null;
+//!     tracer.start(&slot, group, event);
+//!     ... do stuff ...
+//!     tracer.end(&slot, group, event);
+//!
+//! Each slot can be used as many times as you like,
+//! but you must alternate calls to start and end,
+//! and you must end every event.
+//!
+//!     // good
+//!     tracer.start(&slot, group_a, event_a);
+//!     tracer.end(&slot, group_a, event_a);
+//!     tracer.start(&slot, group_b, event_b);
+//!     tracer.end(&slot, group_b, event_b);
+//!
+//!     // bad
+//!     tracer.start(&slot, group_a, event_a);
+//!     tracer.start(&slot, group_b, event_b);
+//!     tracer.end(&slot, group_b, event_b);
+//!     tracer.end(&slot, group_a, event_a);
+//!
+//!     // bad
+//!     tracer.end(&slot, group_a, event_a);
+//!     tracer.start(&slot, group_a, event_a);
+//!
+//!     // bad
+//!     tracer.start(&slot, group_a, event_a);
+//!     std.os.exit(0);
+//!
+//! Before freeing a slot, you should `assert(slot == null)`
+//! to ensure that you didn't forget to end an event.
+//!
+//! Each `Event` has an `EventGroup`.
+//! Within each group, event spans should form a tree.
+//!
+//!     // good
+//!     tracer.start(&a, group, ...);
+//!     tracer.start(&b, group, ...);
+//!     tracer.end(&b, group, ...);
+//!     tracer.end(&a, group, ...);
+//!
+//!     // bad
+//!     tracer.start(&a, group, ...);
+//!     tracer.start(&b, group, ...);
+//!     tracer.end(&a, group, ...);
+//!     tracer.end(&b, group, ...);
+//!
+//! The tracer itself will not object to non-tree spans,
+//! but some backends will either refuse to open the trace or will render it weirdly.
+//!
+//! If you're having trouble making your spans form a tree, feel free to just add new groups.
+
+const std = @import("std");
+const assert = std.debug.assert;
+const Allocator = std.mem.Allocator;
+const log = std.log.scoped(.tracer);
+
+const Time = @import("./time.zig").Time;
+const build_options = @import("tigerbeetle_build_options");
+
+const backend = if (std.meta.eql(build_options.tracer_backend, "none"))
```

Comment:
At the moment tracer_backend is always defined, with the default set in build.zig. 

I *think* we could pass an enum instead of a string (I just reread https://github.com/ziglang/zig/blob/6d44a6222d6eba600deb7f16c124bfa30628fb60//lib/std/build/OptionsStep.zig#L36 - I missed the second switch when I first read it.) We'd have to define TracerBackend in build.zig instead of in tracer, but that's not so bad. What do you think?

Reply:
Actually instead could you move `Backend` and `backend` to `config.zig` (as `TracerBackend` and `tracer_backend`)?

That way all "config" validation is in one place — configuring the backend is similar to configuring the log level, so we can treat it the same way.

Reply:
Still set via build options?

Reply:
Yeah — eventually we will do the same for other config options as well, so that is still good.

Reply:
I put TracerBackend in build.zig to avoid having to import config from build, but the rest is in config now.